File: default/theme/cache-sqlite/template/php/cache.php
<?php

class Cache { // SQLite-based caching system with retry mechanism
	private static $db = null;
	private static $maxRetries = 3;  // Maximum retry attempts
	private static $retryDelay = 100000;  // Retry delay in microseconds (100ms)

	public static function initializeDb() { // Initialize SQLite database connection and schema
		// alternative: initDb()
		if (self::$db === null) {
			$cacheDir = GetDir('cache');
			$dbFile = $cacheDir . '/cache.db';

			try {
				self::$db = new SQLite3($dbFile);
				self::$db->busyTimeout(5000); // 5 second busy timeout

				$attempts = 0;
				$success = false;

				while (!$success && $attempts < self::$maxRetries) {
					try {
						self::$db->exec("
							CREATE TABLE IF NOT EXISTS cache (
								key TEXT PRIMARY KEY,
								value TEXT,
								version TEXT,
								created_at DATETIME DEFAULT CURRENT_TIMESTAMP
							)
						");
						$success = true;
					} catch (Exception $e) {
						$attempts++;
						if ($attempts >= self::$maxRetries) {
							throw $e;
						}
						WriteLog("initializeDb: retry attempt $attempts: " . $e->getMessage());
						usleep(self::$retryDelay);
					}
				}
			} catch (Exception $e) {
				WriteLog("initializeDb: error: " . $e->getMessage());
				die('Cache initialization failed');
			}
		}
		return self::$db;
	} // initializeDb()

	public static function getDb() { // Get database connection, initializing if needed
		return self::initializeDb();
	} // getDb()

	public static function closeDb() { // Close database connection if open
		if (self::$db !== null) {
			self::$db->close();
			self::$db = null;
		}
	} // closeDb()

	public static function retryOperation($callback) { // Execute operation with retry logic
		$attempts = 0;
		$lastError = null;

		while ($attempts < self::$maxRetries) {
			try {
				return $callback();
			} catch (Exception $e) {
				$lastError = $e;
				$attempts++;
				if (stripos($e->getMessage(), 'database is locked') !== false) {
					WriteLog("retryOperation: database locked, attempt $attempts");
					usleep(self::$retryDelay);
					continue;
				}
				throw $e;
			}
		}

		throw $lastError;
	} // retryOperation()
}

function GetCache($cacheName) { // Retrieve value from cache by key
	return Cache::retryOperation(function() use ($cacheName) {
		$db = Cache::getDb();
		$myVersion = GetMyCacheVersion();

		try {
			$stmt = $db->prepare('SELECT value FROM cache WHERE key = :key AND version = :version');
			if (!$stmt) {
				throw new Exception($db->lastErrorMsg());
			}

			$stmt->bindValue(':key', $cacheName, SQLITE3_TEXT);
			$stmt->bindValue(':version', $myVersion, SQLITE3_TEXT);

			$result = $stmt->execute();
			if ($row = $result->fetchArray(SQLITE3_ASSOC)) {
				WriteLog("GetCache: found value for $cacheName");
				return $row['value'];
			}
		} catch (Exception $e) {
			WriteLog("GetCache: error: " . $e->getMessage());
			throw $e;
		}

		WriteLog("GetCache: no value found for $cacheName");
		return '';
	});
} // GetCache()

function PutCache($cacheName, $content) { // Store value in cache with key
	return Cache::retryOperation(function() use ($cacheName, $content) {
		WriteLog("PutCache: storing $cacheName");

		$db = Cache::getDb();
		$myVersion = GetMyCacheVersion();

		try {
			$stmt = $db->prepare('
				INSERT OR REPLACE INTO cache (key, value, version)
				VALUES (:key, :value, :version)
			');

			if (!$stmt) {
				throw new Exception($db->lastErrorMsg());
			}

			$stmt->bindValue(':key', $cacheName, SQLITE3_TEXT);
			$stmt->bindValue(':value', $content, SQLITE3_TEXT);
			$stmt->bindValue(':version', $myVersion, SQLITE3_TEXT);

			return $stmt->execute() !== false;
		} catch (Exception $e) {
			WriteLog("PutCache: error: " . $e->getMessage());
			throw $e;
		}
	});
} // PutCache()

function UnlinkCache($cacheName) { // Remove cache entries matching key pattern
	// alternative: DeleteCache(), RemoveCache()
	return Cache::retryOperation(function() use ($cacheName) {
		WriteLog("UnlinkCache: removing $cacheName");

		$db = Cache::getDb();
		$myVersion = GetMyCacheVersion();

		try {
			$cacheName = str_replace('*', '%', $cacheName);

			$stmt = $db->prepare('DELETE FROM cache WHERE key LIKE :key AND version = :version');
			if (!$stmt) {
				throw new Exception($db->lastErrorMsg());
			}

			$stmt->bindValue(':key', $cacheName, SQLITE3_TEXT);
			$stmt->bindValue(':version', $myVersion, SQLITE3_TEXT);

			return $stmt->execute() !== false;
		} catch (Exception $e) {
			WriteLog("UnlinkCache: error: " . $e->getMessage());
			throw $e;
		}
	});
} // UnlinkCache()

function CacheExists($cacheName) { // Check if cache entry exists
	return Cache::retryOperation(function() use ($cacheName) {
		$db = Cache::getDb();
		$myVersion = GetMyCacheVersion();

		try {
			$stmt = $db->prepare('SELECT 1 FROM cache WHERE key = :key AND version = :version');
			if (!$stmt) {
				throw new Exception($db->lastErrorMsg());
			}

			$stmt->bindValue(':key', $cacheName, SQLITE3_TEXT);
			$stmt->bindValue(':version', $myVersion, SQLITE3_TEXT);

			$result = $stmt->execute();
			return $result->fetchArray() ? 1 : 0;
		} catch (Exception $e) {
			WriteLog("CacheExists: error: " . $e->getMessage());
			throw $e;
		}
	});
} // CacheExists()

function MigrateCache($cacheKey) { // Migrate cache from filesystem to SQLite
	if (!$cacheKey) {
		WriteLog('MigrateCache: warning: missing cache key');
		return false;
	}

	$CACHEPATH = GetDir('cache');
	$cacheVersion = GetMyCacheVersion();
	
	$filePath = "$CACHEPATH/$cacheVersion/$cacheKey";
	
	if (!file_exists($filePath)) {
		WriteLog('MigrateCache: cache file does not exist: ' . $filePath);
		return false;
	}

	// Read content and migrate to SQLite
	$content = GetFile($filePath);
	if ($content) {
		if (PutCache($cacheKey, $content)) {
			unlink($filePath);
			WriteLog('MigrateCache: migrated ' . $cacheKey . ' from filesystem to SQLite');
			return true;
		} else {
			WriteLog('MigrateCache: failed to write to SQLite cache: ' . $cacheKey);
			return false;
		}
	} else {
		WriteLog('MigrateCache: failed to read cache file: ' . $filePath);
		return false;
	}
} // MigrateCache()

register_shutdown_function(function() {
	Cache::closeDb();
});

File: default/theme/cache-sqlite/template/perl/cache.pl
#!/usr/bin/perl -T

# cache.pl
# This is an alternative to the default file-based cache in default/template/perl/cache.pl
# This module provides SQLite-based caching functionality for the application
# It handles cache initialization, storage, retrieval, and management
# The cache is used to store various types of data including avatars and page content
# Functions:
#   InitializeCache() - Sets up SQLite database connection and creates cache table
#   GetCache() - Retrieves cached content by key
#   PutCache() - Stores content in cache with given key
#   UnlinkCache() - Removes cached item
#   ExpireAvatarCache() - Specifically handles expiring avatar caches
#   MigrateCache() - Migrates cache entries from filesystem to SQLite

# TODO: Consider adding:
# - Cache expiration/TTL support
# - Cache size limits and pruning
# - Better error handling and recovery
# - Connection pooling for better concurrency
# - Prepared statement caching
# - Transaction support for batch operations
# - Compression for large values
# - Metrics/monitoring

use strict;
use 5.010;
use warnings;
use utf8;
use DBI;
use DBD::SQLite;

# POTENTIAL BUG: Global $dbh can lead to connection issues in multi-threaded env
# TODO: Consider making this a package variable or OO implementation
my $dbh;

sub InitializeCache { # connects to SQLite db and creates cache table
# sub InitDb {
	state $cacheDir = GetDir('cache');
	my $dbFile = "$cacheDir/cache.db";

	# POTENTIAL BUG: Race condition possible during directory creation
	if (!-d $cacheDir) {
		require File::Path;
		File::Path::make_path($cacheDir) or die "Failed to create cache directory: $!";
	}

	# TODO: Make these configurable
	my $maxRetries = 3;
	my $retryDelay = 0.1; # 100ms
	my $attempts = 0;
	my $connected = 0;

	# POTENTIAL BUG: No timeout on overall connection attempts
	while (!$connected && $attempts < $maxRetries) {
		eval {
			# TODO: Consider adding WAL mode and other performance optimizations
			$dbh = DBI->connect(
				"dbi:SQLite:dbname=$dbFile",
				"",
				"",
				{
					RaiseError => 1,
					AutoCommit => 1,
					sqlite_busy_timeout => 5000
				}
			);
			$connected = 1;
		};
		if ($@) {
			$attempts++;
			if ($attempts >= $maxRetries) {
				WriteLog('InitializeCache: error: failed after $maxRetries = ' . $maxRetries . '; $@ = ' . $@);
				die "Failed to connect after $maxRetries attempts: $@";
			}
			WriteLog('InitializeCache: retry $attempts = ' . $attempts . '; $@ = ' . $@);
			select(undef, undef, undef, $retryDelay);
		}
	}

	# TODO: Consider adding indexes on version and created_at
	$dbh->do("CREATE TABLE IF NOT EXISTS cache (
		key TEXT PRIMARY KEY,
		value TEXT,
		version TEXT,
		created_at DATETIME DEFAULT CURRENT_TIMESTAMP
	)") or die $dbh->errstr;
} # InitializeCache()

sub GetMyCacheVersion { # returns cache version string
# sub GetCacheVer {
	# POTENTIAL BUG: Hard-coded version prevents proper cache invalidation
	my $cacheVersion = 'b'; #todo make this return something else

	state $initialized;
	if (!$initialized) {
		$initialized = 1;
		InitializeCache();
	}

	WriteLog('GetMyCacheVersion: returning $cacheVersion = ' . $cacheVersion);
	return $cacheVersion;
} # GetMyCacheVersion()

sub GetCache { # $cacheName ; retrieves value from cache by key
# sub CacheGet {
	my $cacheName = shift;
	chomp $cacheName;

	WriteLog('GetCache: called with $cacheName = ' . $cacheName);

	# POTENTIAL BUG: Regex allows potentially dangerous path characters
	if ($cacheName =~ m/^([\/[a-z0-9A-Z_.\/]+)$/i) {
		$cacheName = $1;
		WriteLog('GetCache: sanity check passed, $cacheName = ' . $cacheName);
	} else {
		WriteLog('GetCache: warning: sanity check failed, $cacheName = ' . $cacheName);
		WriteLog('GetCache: returning empty string');
		return '';
	}

	state $myVersion = GetMyCacheVersion();
	WriteLog('GetCache: using cache version = ' . $myVersion);

	# POTENTIAL BUG: No check if $dbh is defined/connected
	return '' unless $dbh;

	# TODO: Consider caching prepared statements
	my $sth = $dbh->prepare("SELECT value FROM cache WHERE key = ? AND version = ?");
	$sth->execute($cacheName, $myVersion);
	my $row = $sth->fetchrow_arrayref();

	# POTENTIAL BUG: No distinction between "not found" and "empty value"
	if ($row) {
		WriteLog('GetCache: found value for key ' . $cacheName . ', length = ' . length($row->[0]));
		return $row->[0];
	} else {
		WriteLog('GetCache: no value found for key ' . $cacheName);
		return undef;
	}
} # GetCache()

sub PutCache { # $cacheName, $content ; stores value in cache
# sub CacheSet {
	my $cacheName = shift;
	chomp($cacheName);

	# POTENTIAL BUG: Regex pattern differs from GetCache()
	if ($cacheName =~ m/^([0-9a-zA-Z\/_.]+)$/) {
		$cacheName = $1;
		WriteLog('PutCache: sanity check passed, $cacheName = ' . $cacheName);
	} else {
		WriteLog('PutCache: warning: sanity check failed, $cacheName = ' . $cacheName);
		return '';
	}

	my $content = shift;

	# POTENTIAL BUG: Empty content is allowed but may cause issues
	if (!defined($content)) {
		WriteLog('PutCache: warning: no $content; caller = ' . join(',', caller));
		return 0;
	}

	WriteLog('PutCache: $cacheName = ' . $cacheName . '; content length = ' . length($content));

	state $myVersion = GetMyCacheVersion();

	# POTENTIAL BUG: No check if $dbh is defined/connected
	return 0 unless $dbh;

	# TODO: Consider using transactions for atomicity
	$dbh->do(
		"INSERT OR REPLACE INTO cache (key, value, version) VALUES (?, ?, ?)",
		undef,
		$cacheName,
		$content,
		$myVersion
	) or return 0;

	return 1;
} # PutCache()

sub UnlinkCache { # $cacheName ; removes cache entries matching pattern
# sub DeleteCache {
# sub RemoveCache {
	my $cacheName = shift;
	chomp($cacheName);

	WriteLog('UnlinkCache: $cacheName = ' . $cacheName);

	state $myVersion = GetMyCacheVersion();

	# POTENTIAL BUG: SQL injection possible with unescaped wildcards
	$cacheName =~ s/\*/%/g;

	# POTENTIAL BUG: No check if $dbh is defined/connected
	return unless $dbh;

	$dbh->do(
		"DELETE FROM cache WHERE key LIKE ? AND version = ?",
		undef,
		$cacheName,
		$myVersion
	);
} # UnlinkCache()

sub CacheExists { # $cacheName ; checks if cache key exists
# sub HasCache {
	my $cacheName = shift;
	chomp($cacheName);

	state $myVersion = GetMyCacheVersion();

	# POTENTIAL BUG: No check if $dbh is defined/connected
	return 0 unless $dbh;

	# TODO: Consider using COUNT(*) instead of SELECT 1
	my $sth = $dbh->prepare("SELECT 1 FROM cache WHERE key = ? AND version = ?");
	$sth->execute($cacheName, $myVersion);
	return $sth->fetchrow_array() ? 1 : 0;
} # CacheExists()

sub GetMessageCacheName { # $itemHash ; returns cache key for message
# sub MessageCacheKey {
	my $itemHash = shift;
	chomp($itemHash);

	# POTENTIAL BUG: IsItem() function not defined in this file
	if (!IsItem($itemHash)) {
		WriteLog('GetMessageCacheName: warning: sanity check failed, $itemHash = ' . $itemHash);
		return '';
	}

	return "message/$itemHash";
} # GetMessageCacheName()

sub ExpireAvatarCache { # $key ; removes avatar caches for key
# sub DeleteAvatarCache {
	my $key = shift;

	if (!defined($key) || $key eq '' || !$key) {
		WriteLog('ExpireAvatarCache: warning: invalid $key; caller = ' . join(',', caller));
		return 0;
	}

	# POTENTIAL BUG: IsFingerprint() function not defined in this file
	if (!IsFingerprint($key) && $key ne '*') {
		WriteLog('ExpireAvatarCache: warning: sanity check failed, $key = ' . $key);
		return 0;
	}

	my $themesValue = GetConfig('theme');
	$themesValue =~ s/[\s]+/ /g;
	my @activeThemes = split(' ', $themesValue);
	my $themeName = $activeThemes[0];

	WriteLog('ExpireAvatarCache: $themeName = ' . $themeName . '; $key = ' . $key);

	return UnlinkCache('avatar%/%/' . $key);
} # ExpireAvatarCache()

sub MigrateCache { # $cacheKey ; migrates cache from filesystem to SQLite
	my $cacheKey = shift;
	chomp($cacheKey);

	if (!$cacheKey) {
		WriteLog('MigrateCache: warning: missing cache key');
		return 0;
	}

	state $CACHEPATH = GetDir('cache');
	state $cacheVersion = GetMyCacheVersion();

	my $filePath = "$CACHEPATH/$cacheVersion/$cacheKey";

	# POTENTIAL BUG: Race condition between check and read
	if (!-e $filePath) {
		WriteLog('MigrateCache: cache file does not exist: ' . $filePath);
		return 0;
	}
	# Read content and migrate to SQLite
	my $content = GetFile($filePath);
	if (!defined($content)) {
		WriteLog('MigrateCache: failed to read cache file: ' . $filePath);
		return 0;
	}

	# POTENTIAL BUG: Race condition between write and unlink
	if (PutCache($cacheKey, $content)) {
		unlink($filePath);
		WriteLog('MigrateCache: migrated ' . $cacheKey . ' from filesystem to SQLite');
		return 1;
	} else {
		WriteLog('MigrateCache: failed to write to SQLite cache: ' . $cacheKey);
		return 0;
	}
} # MigrateCache()

END {
	$dbh->disconnect if $dbh;
}

1;
